redash/models/__init__.py

```diff
@@ -138,28 +191,40 @@ def get_by_id(cls, _id):
         return cls.query.filter(cls.id == _id).one()
 
     def delete(self):
+        # Delete the relevant metadata about a data source first.
+        tables = TableMetadata.query.filter(TableMetadata.data_source_id == self.id).options(load_only('id'))
+        ColumnMetadata.query.filter(
+            ColumnMetadata.table_id.in_(tables.subquery())
+        ).delete(synchronize_session=False)
+
+        tables.delete()
+
         Query.query.filter(Query.data_source == self).update(dict(data_source_id=None, latest_query_data_id=None))
         QueryResult.query.filter(QueryResult.data_source == self).delete()
         res = db.session.delete(self)
         db.session.commit()
         return res
 
     def get_schema(self, refresh=False):
```

Comment:
What about the refresh feature? Users should be able to trigger schema refresh.

Reply:
Indeed, I had taken it out since refreshing with samples is slower and I didn't think it was a good idea to do it synchronously. However, I've added back the ability to refresh.